march_data_collector/src/march_data_collector/cp_calculator.py

```diff
@@ -29,13 +29,31 @@ def __init__(self, tf_buffer, foot_link):
         self.marker.scale.z = 0.03
 
         self.g = 9.81  # gravity constant
+        self.buffer_size = 25  # number of com's to use in estimation of velocity
+        self.polyorder = 4  # order of the polynomial in Saviztky-Golay
+
+        self.com_x_buffer = RingBuffer(capacity=self.buffer_size, dtype=float64)
+        self.com_y_buffer = RingBuffer(capacity=self.buffer_size, dtype=float64)
 
     def calculate_cp(self, com_mark):
         current_time = com_mark.header.stamp
-        time_difference = (current_time - self.prev_t).to_sec()
         if current_time is not self.prev_t:
-            x_dot = (com_mark.pose.position.x - self.prev_x) / time_difference
-            y_dot = (com_mark.pose.position.y - self.prev_y) / time_difference
+
+            self.com_x_buffer.append(com_mark.pose.position.x)
+            self.com_y_buffer.append(com_mark.pose.position.y)
+
+            time_difference = (current_time - self.prev_t).to_sec()
+
+            # window length should be odd, greater then poly order and greater or equal to to the buffer size
+            if len(self.com_x_buffer) <= self.polyorder:
+                return self.marker
+
+            window_length = min(self.buffer_size, len(self.com_x_buffer))
```

Comment:
Isn't the definition of a ringbuffer that it cannot be greater than the given buffer size? So I think you can just use
```suggestion
            window_length = len(self.com_x_buffer)
```